File: common/src/main/java/com/andyanika/translator/common/LocalRepository.java
package com.andyanika.translator.common;

import com.andyanika.translator.common.models.LanguageCode;
import com.andyanika.translator.common.models.TranslateDirection;
import com.andyanika.translator.common.models.TranslateResult;
import com.andyanika.translator.common.models.TranslationRequest;
import com.andyanika.translator.common.models.TranslationRowModel;

import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Maybe;
import io.reactivex.Single;

public interface LocalRepository {
    Single<TranslateResult> translate(TranslationRequest request);
    void addTranslation(TranslateResult translateResult);
    Flowable<List<TranslationRowModel>> getHistory();

    Flowable<List<TranslationRowModel>> getFavorites();
    void addFavorites(int wordId);
    void removeFavorite(int wordId);

    TranslateDirection getLanguageDirection();
    void setLanguageDirection(TranslateDirection direction);
    List<LanguageCode> getAvailableLanguages();
}

File: core/src/main/kotlin/core/interfaces/LocalRepository.kt
package core.interfaces

import core.models.*
import kotlinx.coroutines.flow.Flow

interface LocalRepository {
    suspend fun translate(request: TranslateRequest): TranslateResult?
    suspend fun addTranslation(translateResult: TranslateResult)
    fun getHistory(): Flow<List<FavoriteModel>>
    fun getFavorites(): Flow<List<FavoriteModel>>
    suspend fun addFavorites(wordId: Int)
    suspend fun removeFavorite(wordId: Int)
    suspend fun setLanguageDirection(direction: TranslateDirection<LanguageCode>)
    suspend fun getAvailableLanguages(): List<LanguageCode>

    fun observeSrcLanguage(): Flow<LanguageCode>
    fun observeDstLanguage(): Flow<LanguageCode>
    fun getSrcLanguage(): LanguageCode
    fun getDstLanguage(): LanguageCode
}
